File: cachatore/destinatarios.py
"""Override de destinatarios de notificación — espejo Python del helper que
vive en functions/src/comun.ts y whatsapp-bot/src/destinatarios.js.

Lee el doc `META/destinatarios_notificacion` con cache 5 min y devuelve el
DNI override para una key dada, o el `fallback` (típicamente la constante
hardcoded del módulo) si no hay override válido.

El doc lo edita el admin desde la pantalla "Destinatarios de notificación"
en la app. Cambiar el destinatario NO requiere reiniciar el servicio
cachatore: el cache TTL 5 min hace que la próxima ronda tome el valor
nuevo sin downtime.

Si Firestore falla / el doc no existe / la key no tiene override, devuelve
el fallback. Un Firestore caído no rompe el bot — seguimos usando los
valores históricos hardcoded.
"""
import time

import choferes

COL_META = "META"
DOC_DESTINATARIOS = "destinatarios_notificacion"
TTL_SEG = 5 * 60

_cache: dict | None = None
_cache_expira: float = 0.0


def _cargar() -> dict:
    """Lee el doc con cache 5 min. Si falla, devuelve el cache viejo (aun
    vencido) o un dict vacío — los callers ya tienen fallback hardcoded.
    """
    global _cache, _cache_expira
    if _cache is not None and time.monotonic() < _cache_expira:
        return _cache
    try:
        db = choferes._db()
        snap = db.collection(COL_META).document(DOC_DESTINATARIOS).get()
        if snap.exists:
            _cache = snap.to_dict() or {}
        else:
            _cache = {}
        _cache_expira = time.monotonic() + TTL_SEG
        return _cache
    except Exception:
        return _cache or {}


def obtener_destinatario(key: str, fallback: str) -> str:
    """Devuelve el DNI para `key` desde Firestore, o `fallback` si no hay
    override válido."""
    mp = _cargar()
    v = mp.get(key)
    if isinstance(v, str) and v.strip():
        return v.strip()
    return fallback


def invalidar_cache() -> None:
    """Útil para tests."""
    global _cache, _cache_expira
    _cache = None
    _cache_expira = 0.0

File: whatsapp-bot/src/destinatarios.js
// =============================================================================
// DESTINATARIOS DE NOTIFICACIÓN — override desde Firestore
// =============================================================================
//
// Espejo en Node.js del helper que vive en functions/src/comun.ts. Lee el
// doc META/destinatarios_notificacion con cache 5 min y devuelve el DNI
// override para una key dada, o el fallback (típicamente una env var)
// si no hay override válido.
//
// El doc lo edita el admin desde la pantalla "Destinatarios de
// notificación" en la app. Cambiar el destinatario NO requiere reiniciar
// el bot: el cache TTL 5 min hace que el próximo cron tome el valor
// nuevo sin downtime.
//
// Si Firestore falla / el doc no existe / la key no tiene override,
// devuelve el fallback. Esto significa que un Firestore caído no rompe
// el bot — solo seguimos usando los valores históricos del .env.

const admin = require('firebase-admin');

let _db = null;
let _cache = null;
let _cacheExpiraMs = 0;
const TTL_MS = 5 * 60 * 1000;

function _getDb() {
  if (_db) return _db;
  _db = admin.firestore();
  return _db;
}

/**
 * Lee el doc META/destinatarios_notificacion con cache 5 min.
 * Devuelve un map plano { key: dni }. Si la lectura falla, devuelve
 * el cache vencido (si existe) o un map vacío — los callers ya tienen
 * fallback al env var, así que Firestore caído no rompe nada.
 */
async function _cargar() {
  if (_cache && Date.now() < _cacheExpiraMs) return _cache;
  try {
    const snap = await _getDb()
      .collection('META')
      .doc('destinatarios_notificacion')
      .get();
    if (snap.exists) {
      _cache = snap.data() || {};
      _cacheExpiraMs = Date.now() + TTL_MS;
      return _cache;
    }
    // Doc no existe — cacheamos vacío para no leer Firestore en cada
    // llamada subsiguiente durante 5 min.
    _cache = {};
    _cacheExpiraMs = Date.now() + TTL_MS;
    return _cache;
  } catch (e) {
    // Fallback al cache viejo (aun vencido) o map vacío.
    return _cache || {};
  }
}

/**
 * Devuelve el DNI para `key` desde Firestore, o `fallback` si no hay
 * override válido. `fallback` típicamente es una env var (ej.
 * `process.env.SERVICE_DESTINATARIO_DNI`) o un string hardcoded.
 *
 * Keys conocidas:
 *   - serviceDiario, vencimientosProximosConsolidado (bot crons)
 *   - mantenimientoBot, driftsAsignaciones, parteMantenimientoVolvo,
 *     excesosJornada, conductaManejo, bypassSeguridad (CFs)
 *   - cachatoreEncargado (cachatore Python)
 *   - colaCreciente (bot health alert)
 */
async function obtenerDestinatario(key, fallback) {
  const map = await _cargar();
  const v = map[key];
  if (typeof v === 'string' && v.trim().length > 0) return v.trim();
  return fallback;
}

/** Invalida el cache. Útil para tests. */
function invalidarCache() {
  _cache = null;
  _cacheExpiraMs = 0;
}

/** Solo para tests: inyecta un db mock (evita admin.firestore() real). */
function _setDbParaTests(db) {
  _db = db;
}

module.exports = {
  obtenerDestinatario,
  invalidarCache,
  _setDbParaTests,
};
